game-core/src/main/java/games/strategy/engine/framework/startup/ui/posted/game/pbem/PbemSetupPanel.java

```diff
@@ -0,0 +1,245 @@
+package games.strategy.engine.framework.startup.ui.posted.game.pbem;
+
+import com.google.common.base.Preconditions;
+import games.strategy.engine.data.GameData;
+import games.strategy.engine.framework.message.PlayerListing;
+import games.strategy.engine.framework.startup.launcher.ILauncher;
+import games.strategy.engine.framework.startup.launcher.LocalLauncher;
+import games.strategy.engine.framework.startup.mc.GameSelectorModel;
+import games.strategy.engine.framework.startup.mc.HeadedLaunchAction;
+import games.strategy.engine.framework.startup.ui.PlayerSelectorRow;
+import games.strategy.engine.framework.startup.ui.PlayerType;
+import games.strategy.engine.framework.startup.ui.SetupPanel;
+import games.strategy.engine.framework.startup.ui.posted.game.DiceServerEditor;
+import games.strategy.engine.random.PbemDiceRoller;
+import java.awt.Dialog;
+import java.awt.Dimension;
+import java.awt.GridBagLayout;
+import java.awt.Window;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Observable;
+import java.util.Observer;
+import java.util.Optional;
+import javax.swing.Action;
+import javax.swing.JButton;
+import javax.swing.JOptionPane;
+import javax.swing.JPanel;
+import javax.swing.JScrollPane;
+import javax.swing.JTabbedPane;
+import javax.swing.SwingUtilities;
+import javax.swing.border.EmptyBorder;
+import org.triplea.swing.SwingAction;
+import org.triplea.swing.jpanel.GridBagConstraintsAnchor;
+import org.triplea.swing.jpanel.GridBagConstraintsBuilder;
+import org.triplea.swing.jpanel.GridBagConstraintsFill;
+
+/**
+ * A panel for setting up Play by Email/Forum. This panel listens to the GameSelectionModel so it
+ * can refresh when a new game is selected or save game loaded The MainPanel also listens to this
+ * panel, and we notify it through the notifyObservers()
+ */
+public class PbemSetupPanel extends SetupPanel implements Observer {
```

Comment:
While I appreciate the split between those 2 classes, (I believe at some point it was even possible to play both PbEM and PbF at the same time 🤷‍♂) I think simply copying over 80% of the code isn't really something I'd consider clean code.

Did you consider creating an abstract `PlayByX` class that implements the common functionality and leaves the details for each subclass to decide, simply so we don't end up with that much duplicate code.

Reply:
To respond directly to this now that I have a few more moments, the additional comments section addressed this concern directly:
> PbemSetupPanel is gratuitously duplicated. This duplication should not cause too many issues as we will be looking to replace such classes, and/or we'll have better opportunity to simplify a purpose specific panel rather than one that needs to be abstract. The goal of this update is more to accomplish the screen and button split before further updates take place (placing more value here on accomplishing the UX change before we do more under-the-hood work and enabling a better path for simplifications).

> Did you consider creating an abstract PlayByX class that implements the common functionality and leaves the details for each subclass to decide, simply so we don't end up with that much duplicate code.

Yes, an abstract class based around what we have now could easily lead to more complexity, or alternatively we might invest in cleaning up UI code that has many issues when we are looking to replace/delete it entirely. 

I'd note, that DRY does not mean to do human pattern matching on code and eliminate all common patterns, but to condense ideas to be expressed in one place. For the perspective of a single screen, having similar if not duplicated UI code, given we have only two instances, and the only thing they share thus far is dice roller, the duplication is actually not that bad. 

Second, the code smell is just a smell right now, I don't think this will actually cause any problems or missed updates where an update that shoudl be applied to both instances would actually only be used in one.

Third point regarding the duplication, it could easily be better to inline the PBF and PBEM editor and then clean-up/refactor from there and leave only the dice editor as the common and extracted code.